File: backend/app/arena/judging.py
"""自动判分器 — 任务模板的 L1 格式/约束验证（无需 LLM）。

判分类型（``JudgeSpec.type``）：
- ``keyword``: 答案须命中指定关键词（any_of / all_of / min_hits）
- ``json``: 答案可解析为 JSON，且满足 required_fields / field_types
- ``code``: 提取 Python 代码块（或全文），AST 语法校验 + 可选 must_contain
- ``numeric``: 从答案提取数字并按 operator/value/tolerance 比较
- ``exclude``: 答案不得包含指定模式（如"无法回答"）
- ``regex``: 答案匹配正则

设计原则：所有判分器都是纯函数、确定性、可测试；不做语义判断
（语义验证属于 Harness verify 的 L3 层，由 LLM 完成）。
"""

from __future__ import annotations

import ast
import json
import re
from typing import Literal

from pydantic import BaseModel, Field

# 判分类型
JudgeType = Literal["keyword", "json", "code", "numeric", "exclude", "regex", "none"]

# 提取 Python 代码块：```python ... ``` 或 ``` ... ```
_CODE_FENCE = re.compile(r"```(?:python|py)?\s*\n(.*?)\n?\s*```", re.DOTALL)


class JudgeSpec(BaseModel):
    """模板判分规则。"""

    type: JudgeType
    # keyword
    any_of: list[str] = Field(default_factory=list)
    all_of: list[str] = Field(default_factory=list)
    min_hits: int = Field(default=1, ge=1)
    # json
    required_fields: list[str] = Field(default_factory=list)
    # code
    must_contain: list[str] = Field(default_factory=list)
    max_len: int = Field(default=8000, ge=1)
    # numeric
    operator: Literal["==", ">=", "<=", ">", "<"] = "=="
    value: float = 0.0
    tolerance: float = Field(default=0.0, ge=0.0)
    # exclude / regex
    patterns: list[str] = Field(default_factory=list)
    pattern: str = ""


class JudgeResult(BaseModel):
    """单条答案的判分结果。"""

    passed: bool
    reason: str
    details: list[str] = Field(default_factory=list)


def _extract_numbers(text: str) -> list[float]:
    """提取文本中所有数字（整数/小数，支持千分位与负号）。"""
    return [float(m) for m in re.findall(r"-?\d+(?:,\d{3})*(?:\.\d+)?", text.replace(",", ""))]


def _extract_code(text: str) -> str:
    """优先提取 fenced 代码块；无 fence 时返回全文（去掉可能的 md 标记）。"""
    m = _CODE_FENCE.search(text)
    if m:
        return m.group(1).strip()
    return text.strip()


def _check_numeric(answer: str, spec: JudgeSpec) -> JudgeResult:
    nums = _extract_numbers(answer)
    if not nums:
        return JudgeResult(
            passed=False,
            reason="答案中未提取到数字",
            details=[f"期望 {spec.operator} {spec.value}"],
        )
    got = nums[0]
    target = spec.value
    if spec.operator == "==":
        ok = abs(got - target) <= spec.tolerance
    elif spec.operator == ">=":
        ok = got >= target
    elif spec.operator == "<=":
        ok = got <= target
    elif spec.operator == ">":
        ok = got > target
    else:
        ok = got < target
    return JudgeResult(
        passed=ok,
        reason="数字比较通过" if ok else f"数字比较失败: {got} {spec.operator} {target}",
        details=[f"提取数字: {got}", f"期望: {spec.operator} {target}"],
    )


def _check_json(answer: str, spec: JudgeSpec) -> JudgeResult:
    # 尝试直接解析；失败则提取第一个 {...} 或 [...] 块
    candidates = [answer.strip()]
    brace = re.search(r"[{\[].*[}\]]", answer, re.DOTALL)
    if brace:
        candidates.append(brace.group(0))
    for cand in candidates:
        try:
            data = json.loads(cand)
            break
        except (json.JSONDecodeError, ValueError):
            continue
    else:
        return JudgeResult(passed=False, reason="答案不是合法 JSON", details=["JSON 解析失败"])
    if not isinstance(data, dict):
        return JudgeResult(passed=False, reason="JSON 顶层应为对象", details=[f"实际类型: {type(data).__name__}"])
    missing = [f for f in spec.required_fields if f not in data]
    if missing:
        return JudgeResult(passed=False, reason=f"缺少字段: {', '.join(missing)}", details=missing)
    return JudgeResult(passed=True, reason="JSON 解析与字段校验通过", details=list(data.keys()))


def _check_code(answer: str, spec: JudgeSpec) -> JudgeResult:
    code = _extract_code(answer)
    if len(code) > spec.max_len:
        return JudgeResult(passed=False, reason=f"代码过长（{len(code)} > {spec.max_len}）", details=["长度超限"])
    missing = [kw for kw in spec.must_contain if kw not in code]
    if missing:
        return JudgeResult(passed=False, reason=f"缺少关键字: {', '.join(missing)}", details=missing)
    try:
        ast.parse(code)
    except SyntaxError as exc:
        return JudgeResult(
            passed=False,
            reason=f"Python 语法错误: {exc.msg}",
            details=[f"行 {exc.lineno}"],
        )
    return JudgeResult(passed=True, reason="Python 语法校验通过", details=[f"{len(code)} 字符"])


def _check_keyword(answer: str, spec: JudgeSpec) -> JudgeResult:
    hits: list[str] = []
    for kw in spec.any_of:
        if kw in answer:
            hits.append(kw)
    for kw in spec.all_of:
        if kw in answer:
            hits.append(kw)
        else:
            return JudgeResult(passed=False, reason=f"缺少关键词: {kw}", details=[f"需包含: {kw}"])
    if len(hits) < spec.min_hits:
        return JudgeResult(
            passed=False,
            reason=f"关键词命中不足（{len(hits)}/{spec.min_hits}）",
            details=[f"命中的关键词: {hits or '无'}"],
        )
    return JudgeResult(passed=True, reason=f"关键词命中 {len(hits)} 个", details=hits)


def _check_exclude(answer: str, spec: JudgeSpec) -> JudgeResult:
    found = [p for p in spec.patterns if p in answer]
    if found:
        return JudgeResult(passed=False, reason=f"答案包含禁用模式: {', '.join(found)}", details=found)
    return JudgeResult(passed=True, reason="未包含禁用模式", details=[])


def _check_regex(answer: str, spec: JudgeSpec) -> JudgeResult:
    if not spec.pattern:
        return JudgeResult(passed=False, reason="未配置正则", details=[])
    # 防御 ReDoS：截断超长输入（预置模板最长匹配不超过 100 字符）
    safe_answer = answer[:2000]
    try:
        ok = re.search(spec.pattern, safe_answer) is not None
    except re.error as exc:
        return JudgeResult(passed=False, reason=f"正则无效: {exc}", details=[])
    return JudgeResult(
        passed=ok,
        reason="正则匹配通过" if ok else "正则未匹配",
        details=[spec.pattern],
    )


def judge_answer(answer: str, spec: JudgeSpec) -> JudgeResult:
    """按 JudgeSpec 判分单条答案。空答案直接失败（``none`` 类型除外）。"""
    if spec.type == "none":
        return JudgeResult(passed=True, reason="该模板不支持自动判分", details=[])
    if not answer or not answer.strip():
        return JudgeResult(passed=False, reason="答案为空", details=[])
    if spec.type == "keyword":
        return _check_keyword(answer, spec)
    if spec.type == "json":
        return _check_json(answer, spec)
    if spec.type == "code":
        return _check_code(answer, spec)
    if spec.type == "numeric":
        return _check_numeric(answer, spec)
    if spec.type == "exclude":
        return _check_exclude(answer, spec)
    if spec.type == "regex":
        return _check_regex(answer, spec)
    return JudgeResult(passed=False, reason=f"未知判分类型: {spec.type}", details=[])


def judge_answers(answers: dict[str, str], spec: JudgeSpec) -> dict[str, JudgeResult]:
    """批量判分：{label: answer} → {label: JudgeResult}。"""
    return {label: judge_answer(text, spec) for label, text in answers.items()}

File: packages/evaluation/evaluation/src/judging.ts
/**
 * @file judging
 * @description Deterministic L1 auto-judging for task templates plus LLM-as-Judge.
 *
 * Responsibilities:
 * - Validate answer format and constraints per deterministic judge type
 * - Support keyword, json, code, numeric, exclude, regex, none, and llm types
 * - Offer async LLM judging behind an injected text-invoke port (fail-closed)
 */

import { sanitizeErrorMessage, type JudgeResult, type JudgeSpec, type LlmJudgeAdapter, type LlmJudgeContext } from "@agentprism/contracts";
import { detectInjection, sanitizeForJson } from "@agentprism/contracts";
import { llmJudgePrompt as buildLlmJudgePrompt } from "./prompts.js";

export { buildLlmJudgePrompt };

// The info string (language tag) after the opening fence is stripped for any
// language (```go, ```python3.11, a bare ```), not just python: a tagged fence
// that failed to strip would push the raw ``` markers into max_len and keyword checks.
const CODE_FENCE = /```[^\n`]*\n([\s\S]*?)\n?\s*```/;

function extractNumbers(text: string): number[] {
  const cleaned = text.replaceAll(",", "");
  const matches = cleaned.match(/-?\d+(?:\.\d+)?/g) ?? [];
  return matches.map((m) => Number.parseFloat(m));
}

function extractCode(text: string): string {
  const match = CODE_FENCE.exec(text);
  if (match?.[1] !== undefined) return match[1].trim();
  return text.trim();
}

function result(passed: boolean, reason: string, details: string[] = []): JudgeResult {
  return { passed, reason, details };
}

function checkNumeric(answer: string, spec: JudgeSpec): JudgeResult {
  const numbers = extractNumbers(answer);
  const got = numbers[0];
  if (got === undefined) {
    return result(false, "No number extracted from answer", [`Expected ${spec.operator} ${spec.value}`]);
  }
  const target = spec.value;
  let ok: boolean;
  if (spec.operator === "==") ok = Math.abs(got - target) <= spec.tolerance;
  else if (spec.operator === ">=") ok = got >= target;
  else if (spec.operator === "<=") ok = got <= target;
  else if (spec.operator === ">") ok = got > target;
  else ok = got < target;
  return result(
    ok,
    ok ? "Numeric comparison passed" : `Numeric comparison failed: ${got} ${spec.operator} ${target}`,
    [`Extracted number: ${got}`, `Expected: ${spec.operator} ${target}`],
  );
}

function checkJson(answer: string, spec: JudgeSpec): JudgeResult {
  const candidates = [answer.trim()];
  const brace = /[{[][\s\S]*[}\]]/.exec(answer);
  if (brace !== null) candidates.push(brace[0]);
  let data: unknown;
  let parsed = false;
  for (const candidate of candidates) {
    try {
      data = JSON.parse(candidate);
      parsed = true;
      break;
    } catch {
      continue;
    }
  }
  if (!parsed) {
    return result(false, "Answer is not valid JSON", ["JSON parse failed"]);
  }
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    return result(false, "JSON top-level value must be an object", [`Actual type: ${data === null ? "null" : typeof data}`]);
  }
  const record = data as Record<string, unknown>;
  // Own-property check: `in` walks the prototype chain, so a required field named
  // "toString"/"constructor" would falsely pass without existing in the answer.
  const missing = spec.required_fields.filter((field) => !Object.hasOwn(record, field));
  if (missing.length > 0) {
    return result(false, `Missing fields: ${missing.join(", ")}`, missing);
  }
  return result(true, "JSON parse and field checks passed", Object.keys(record));
}

function checkCode(answer: string, spec: JudgeSpec): JudgeResult {
  const code = extractCode(answer);
  if (code.length > spec.max_len) {
    return result(false, `Code too long (${code.length} > ${spec.max_len})`, ["Length exceeded"]);
  }
  const missing = spec.must_contain.filter((keyword) => !code.includes(keyword));
  if (missing.length > 0) {
    return result(false, `Missing keywords: ${missing.join(", ")}`, missing);
  }
  return result(true, "Code block checks passed", [`${code.length} chars`]);
}

function checkKeyword(answer: string, spec: JudgeSpec): JudgeResult {
  // Deduplicated counting: a keyword present in both any_of/all_of counts once.
  const hits = new Set<string>();
  for (const keyword of spec.any_of) {
    if (answer.includes(keyword)) hits.add(keyword);
  }
  for (const keyword of spec.all_of) {
    if (answer.includes(keyword)) {
      hits.add(keyword);
    } else {
      return result(false, `Missing keyword: ${keyword}`, [`Must include: ${keyword}`]);
    }
  }
  if (hits.size < spec.min_hits) {
    return result(false, `Not enough keyword hits (${hits.size}/${spec.min_hits})`, [
      `Hit keywords: ${hits.size > 0 ? [...hits].join(", ") : "none"}`,
    ]);
  }
  return result(true, `Keyword hits: ${hits.size}`, [...hits]);
}

function checkExclude(answer: string, spec: JudgeSpec): JudgeResult {
  const found = spec.patterns.filter((pattern) => answer.includes(pattern));
  if (found.length > 0) {
    return result(false, `Answer contains excluded patterns: ${found.join(", ")}`, found);
  }
  return result(true, "No excluded patterns found", []);
}

function checkRegex(answer: string, spec: JudgeSpec): JudgeResult {
  if (spec.pattern === "") {
    return result(false, "Regex not configured", []);
  }
  if (spec.pattern.length > 500) {
    return result(false, "Regex too long; rejected", []);
  }
  // ReDoS defense: truncate over-long input (nested-quantifier backtracking risk remains; templates are internal config)
  const safeAnswer = answer.slice(0, 2000);
  let ok: boolean;
  try {
    ok = new RegExp(spec.pattern).test(safeAnswer);
  } catch (error) {
    return result(false, `Invalid regex: ${error instanceof Error ? error.message : "Error"}`, []);
  }
  return result(ok, ok ? "Regex matched" : "Regex did not match", [spec.pattern]);
}

/** Judges one answer by JudgeSpec. Empty answers fail outright (except the none type). */
function judgeAnswer(answer: string, spec: JudgeSpec): JudgeResult {
  if (spec.type === "none") {
    return result(true, "This template does not support auto-judging");
  }
  if (spec.type === "llm") {
    return result(false, "LLM judge requires async judging (use judgeAnswersAsync)");
  }
  if (answer.trim() === "") {
    return result(false, "Answer is empty");
  }
  switch (spec.type) {
    case "keyword":
      return checkKeyword(answer, spec);
    case "json":
      return checkJson(answer, spec);
    case "code":
      return checkCode(answer, spec);
    case "numeric":
      return checkNumeric(answer, spec);
    case "exclude":
      return checkExclude(answer, spec);
    case "regex":
      return checkRegex(answer, spec);
    default:
      return result(false, `Unknown judge type: ${spec.type}`);
  }
}

/**
 * Parses one LLM judge response fail-closed (parse failures count as failing).
 * Deliberately no lenient substring fallback: an answer smuggling
 * `"passed": true` outside valid JSON must not earn a pass.
 */
export function parseLlmJudgeResponse(text: string, spec: JudgeSpec): JudgeResult {
  let parsed: Record<string, unknown>;
  try {
    parsed = JSON.parse(sanitizeForJson(text)) as Record<string, unknown>;
  } catch {
    return result(false, "LLM judge parse failed; treated as not passed");
  }
  const scoreRaw = parsed.score;
  const score = typeof scoreRaw === "number" && Number.isFinite(scoreRaw) ? Math.min(1, Math.max(0, scoreRaw)) : null;
  const reason = typeof parsed.reason === "string" && parsed.reason !== "" ? parsed.reason.slice(0, 500) : "LLM judge verdict";
  if (score !== null) {
    const passed = score >= spec.passing_score;
    return result(passed, reason, [`score=${score}`, `threshold=${spec.passing_score}`]);
  }
  return result(parsed.passed === true, reason);
}

/** Judges one answer with the LLM adapter (fail-closed on adapter/parse errors). */
async function judgeAnswerLlm(
  answer: string,
  spec: JudgeSpec,
  adapter: LlmJudgeAdapter,
  context: LlmJudgeContext = {},
): Promise<JudgeResult> {
  if (answer.trim() === "") return result(false, "Answer is empty");
  // The answer rides inside the judge prompt: a model under test smuggling
  // override prose fails here instead of steering the verdict (harness parity).
  if (detectInjection(answer)) {
    return result(false, "Possible prompt injection detected; treated as not passed");
  }
  try {
    const response = await adapter.invoke(buildLlmJudgePrompt(answer, spec, context.question));
    return parseLlmJudgeResponse(response, spec);
  } catch (error) {
    return result(false, `LLM judge failed; treated as not passed: ${sanitizeErrorMessage(error)}`);
  }
}

/** Batch judging: {label: answer} → {label: JudgeResult} (pure deterministic rules, no model).
 *
 * @param answers Column label to final-answer text.
 * @param spec Judge spec selecting the rule (keyword/json/code/numeric/exclude/regex/none/llm).
 * @returns Label to verdict; empty answers fail outright (except the none type).
 *   llm specs fail closed here (use judgeAnswersAsync with a judge model).
 */
export function judgeAnswers(answers: Record<string, string>, spec: JudgeSpec): Record<string, JudgeResult> {
  const out: Record<string, JudgeResult> = {};
  for (const [label, text] of Object.entries(answers)) {
    out[label] = judgeAnswer(text, spec);
  }
  return out;
}

/** Async batch judging: deterministic types stay sync; llm type calls the adapter per answer.
 *
 * Answers judge sequentially (not Promise.all): the judge model is a shared
 * downstream, and one slow verdict must not fan out into model overload.
 *
 * @param answers Column label to final-answer text.
 * @param spec Judge spec (llm uses the adapter + optional question context).
 * @param adapter Host-supplied LLM text-invoke port (ignored for deterministic types).
 * @param context Optional question context for the judge prompt.
 * @returns Label to verdict; adapter/parse failures fail closed per answer, never throw.
 */
export async function judgeAnswersAsync(
  answers: Record<string, string>,
  spec: JudgeSpec,
  adapter: LlmJudgeAdapter,
  context: LlmJudgeContext = {},
): Promise<Record<string, JudgeResult>> {
  if (spec.type !== "llm") return judgeAnswers(answers, spec);
  const out: Record<string, JudgeResult> = {};
  for (const [label, text] of Object.entries(answers)) {
    out[label] = await judgeAnswerLlm(text, spec, adapter, context);
  }
  return out;
}
